Added WeightedCategorical-Prior

JasperMartins opened the pull request:
This pull request adds a `WeightedCategorical`-prior that subsumes the existing `Categorical`-Prior. Functionally, the only difference between the two is that `WeightedCategorical` supports the setting of relative weights for all categories.

mj-will commented:
Hi @JasperMartins, thank for implementing this and apologies for the delay in looking over it.

The changes look good to me but following a recent change to the `Categorical` prior (see https://github.com/bilby-dev/bilby/pull/947) I think this will need slightly reworking to address the conflicts. Do you think you'd have time/capacity to do this?

mj-will commented on bilby/core/prior/analytical.py:
```diff
@@ -1482,8 +1509,31 @@ def rescale(self, val):
+                return 0
+        else:
+            index = np.searchsorted(self.values, val)
+            index[index == self.nvalues] = self.nvalues - 1
```
Could you help me understand why this is needed?

I understand that if `searchsorted` is given a value greater than the last entry in `values`, then it will return the last index+1. When will this happen? Since this is discrete I thought the values should always match exactly and that seems to give the right index, e.g.

````python
>>> np.searchsorted([1, 3, 5], 5)
np.int64(2)
```

JasperMartins replied:
If `val` is not in `self.values` (and greater than the last value), it can happen, and then the next line (`self.values[index]`) would fail in an obtuse way.
A more elegant solution would probably be to use `np.where` to check for equality from the start.
This actually made me wonder, though, whether returning 0 is actually the desired behavior of the CDF if `val` is not strictly in `self.values`.

mj-will reviewed (commented):
LGTM, just one question from my side.

JasperMartins commented:
Hi @ColmTalbot,

Sorry for the delay, I had been on vacation.
So, prompted by the init-from-repr issue, I have added the new priors to the general assembly of prior-tests, to catch other issues, which also uncovered a small bug in the cdf function for values below the minimum value in the prior-support.
I have fixed these issues and brought the branch up to date.

In principle, the tests in the file `analytical_test.py` could now probably be dropped.
